mentoring/answer.py

```diff
@@ -24,39 +21,146 @@
+            name = self.name
+        # Consistency check - we should have a name by now
+        if not name:
+            raise ValueError('AnswerBlock.name field need to be set to a non-null/empty value')
+
+        student_id = self._get_student_id()
+        course_id = self._get_course_id()
+
+        answer_data, _ = Answer.objects.get_or_create(
+            student_id=student_id,
+            course_id=course_id,
+            name=name,
+        )
+        return answer_data
+
+    def validate_field_data(self, validation, data):
+        """
+        Validate this block's field data.
+        """
+        super(AnswerMixin, self).validate_field_data(validation, data)
+
+        def add_error(msg):
+            validation.add(ValidationMessage(ValidationMessage.ERROR, msg))
+
+        if not data.name:
+            add_error(u"A Question ID is required.")
+
+    def _(self, text):
+        """ translate text """
+        return self.runtime.service(self, "i18n").ugettext(text)
+
+
+@XBlock.needs("i18n")
+class AnswerBlock(AnswerMixin, StepMixin, StudioEditableXBlockMixin, XBlock):
     """
     A field where the student enters an answer
 
     Must be included as a child of a mentoring block. Answers are persisted as django model instances
     to make them searchable and referenceable across xblocks.
     """
-    read_only = Boolean(help="Display as a read-only field", default=False, scope=Scope.content)
-    default_from = String(help="If specified, the name of the answer to get the default value from",
-                          default=None, scope=Scope.content)
+    name = String(
+        display_name=_("Question ID (name)"),
+        help=_("The ID of this block. Should be unique unless you want the answer to be used in multiple places."),
+        default="",
+        scope=Scope.content
+    )
+    default_from = String(
+        display_name=_("Default From"),
+        help=_("If a question ID is specified, get the default value from this answer."),
+        default=None,
+        scope=Scope.content
+    )
+    min_characters = Integer(
+        display_name=_("Min. Allowed Characters"),
+        help=_("Minimum number of characters allowed for the answer"),
+        default=0,
+        scope=Scope.content
+    )
+    question = String(
+        display_name=_("Question"),
+        help=_("Question to ask the student"),
+        scope=Scope.content,
+        default=""
+    )
+    weight = Float(
+        display_name=_("Weight"),
+        help=_("Defines the maximum total grade of the answer block."),
+        default=1,
+        scope=Scope.settings,
+        enforce_type=True
+    )
+
+    editable_fields = ('question', 'name', 'min_characters', 'weight', 'default_from')
+
+    @property
+    def studio_display_name(self):
+        if not self.lonely_step:
+            return self._(u"Question {number}").format(number=self.step_number)
+        return self._(u"Question")
+
+    def __getattribute__(self, name):
```

Comment:
@cpennington 

The outcome I'm trying to get is that the title of this block as seen in Studio is a dynamically generated string ("Question X") based on this block's position among its parent's children. So it must be an instance method of some sort.

All I'm trying to avoid are errors in Studio and Studio displaying strings other than the string I want used as the title.

If I just do `@property def display_name(self):`, it causes errors in Studio, which tries to access the property as though it were a field, and then gives `AttributeError: display_name.default does not exist`. I think a mixin injected by the edX runtime has already contributed a default `display_name` field, so Studio tries to treat this property as a field.

This seems to be what `display_name_with_default` is designed for, but if I implement that, Studio sometimes doesn't bother to read its value, and instead just display's the block type as the display_name in some cases. The following two screenshots show the same pb-mcq block on two different pages:

Correctly checks display_name_with_default:
![screen shot 2015-03-26 at 9 48 59 am](https://cloud.githubusercontent.com/assets/945577/6851641/9a38fc7e-d39d-11e4-800d-bcdac2fd5f7f.png)

Wrong:
![screen shot 2015-03-26 at 9 49 30 am](https://cloud.githubusercontent.com/assets/945577/6851658/a60a44a4-d39d-11e4-8748-90a2d115a9fd.png)


At this point, though, I think I should just use this approach and open a separate PR to make Studio's behavior consistent.